# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/webauthn/WebAuthnManager.h
# Commit: f7937d3264db
# Full Hash: f7937d3264db00771b46cb1fcba71640d8df05cb
# Author: J.C. Jones <jjones@mozilla.com>
# Date: 2019-03-30 09:32:04
# Regressor Bug: 1448408
# File Overlap Count: 2
# Description:
#   Bug 1448408 - Web Authentication - Don't immediately abort on visibility events r=keeler
#   
#   The published recommendation of L1 for WebAuthn changed the visibility/focus
#   listening behaviors to a SHOULD [1], and Chromium, for reasons like our SoftU2F
#   bug [0], opted to not interrupt on tabswitch/visibility change.
# ==============================================================================

diff -r a5c4f7e6069a -r f7937d3264db dom/webauthn/WebAuthnManager.h
--- a/dom/webauthn/WebAuthnManager.h	Fri Mar 29 19:38:53 2019 +0000
+++ b/dom/webauthn/WebAuthnManager.h	Fri Mar 29 17:59:08 2019 +0000
@@ -48,7 +48,7 @@
 class WebAuthnTransaction {
  public:
   explicit WebAuthnTransaction(const RefPtr<Promise>& aPromise)
-      : mPromise(aPromise), mId(NextId()) {
+      : mPromise(aPromise), mId(NextId()), mVisibilityChanged(false) {
     MOZ_ASSERT(mId > 0);
   }
 
@@ -58,6 +58,10 @@
   // Unique transaction id.
   uint64_t mId;
 
+  // Whether or not visibility has changed for the window during this
+  // transaction
+  bool mVisibilityChanged;
+
  private:
   // Generates a unique id for new transactions. This doesn't have to be unique
   // forever, it's sufficient to differentiate between temporally close
@@ -105,7 +109,9 @@
  protected:
   // Cancels the current transaction (by sending a Cancel message to the
   // parent) and rejects it by calling RejectTransaction().
-  void CancelTransaction(const nsresult& aError) override;
+  void CancelTransaction(const nsresult& aError);
+  // Upon a visibility change, makes note of it in the current transaction.
+  void HandleVisibilityChange() override;
 
  private:
   virtual ~WebAuthnManager();